src/ray/object_manager/pull_manager.h

```diff
@@ -0,0 +1,274 @@
+///
+/// Each pull has a timer, which operates as follows:
+/// 1. The timer begins as soon as the pull object is created.
+/// 2. Whenever we request the object from a new object manager, the timer is
+///    reset.
+/// 3. Whenever we receive a chunk from the object manager that we are
+///    receiving the object from, the timer is reset.
+/// 4. The timer is canceled when the object has been fully received
+///    successfully.
+/// 5. The timer is canceled when the pull lifetime ends.
+/// 6. When the timer expires, we do the following:
+///    - If no remote object managers have the object, we reset the timer.
+///    - If other remote object managers have the object and have not started
+///    - receiving the object, we issue some new requests for the object and
+///      possibly cancel some requests.
+///    - If we are in the process of receiving the object, we cancel the
+///      request and issue at least one new request for the object.
+///    - If the object is required, reset the timer.
+///
+/// When we are pulling an object, we only will read chunks from the first
+/// object manager that sends us a push request. Chunks from other object
+/// managers will be ignored. If we fail to read a chunk from the object
+/// manager that we are not ignoring, then we consider ourselves to be no
+/// longer reading from that object manager.
+///
+/// When do we abort an object creation? We only ever abort a creation for
+/// objects that we are not required to pull (e.g., those that were being
+/// pushed to us or whose pulls were canceled). There is the danger that we
+/// abort an object creation only to have a subsequent read try to recreate
+/// it. To deal with this, we have the main thread abort object creations only
+/// when the lifetime of a pull ends.
+class PullManager {
+ public:
+  /// Construct a PullManager.
+  ///
+  /// \param main_service The service to use for running timer callbacks.
+  /// \param client_id The ID of this object manager.
+  /// \param callback The callback that the pull manager can use to request
+  /// objects, cancel requests, and abort object creations.
+  PullManager(const ClientID &client_id);
+
+  PullManager(const PullManager &other) = delete;
+
+  PullManager &operator=(const PullManager &other) = delete;
+
+  /// Pull an object. This will guarantee that the object is pulled. That is,
+  /// it will keep trying to request the object until CancelPullObject is
+  /// called. CancelPullObject is also called when the object appears locally.
+  ///
+  /// \param object_id The ID of the object to pull.
+  /// \param[out] subscribe_to_locations This will be set to true if the caller
+  /// should subscribe to the object locations and false otherwise.
+  /// \param[out] start_timer This will be set to true if a timer should be
+  /// started for this object ID and false otherwise.
+  /// \return Void.
+  void PullObject(const ObjectID &object_id, bool *subscribe_to_locations,
```

Comment:
Can we handle client cancellation and unsubscribe from locations internally in PullManager? Exposing them to the public API weakens the abstraction provided here.

Same question for timer, and for other API methods here and below.